server/src/main/java/org/apache/druid/server/coordinator/ServerHolder.java

```diff
@@ -155,128 +211,161 @@ public long getAvailableSize()
+
+    updateQueuedSegments(segment, simplify(action), true);
     return true;
   }
 
-  public boolean isServingSegment(SegmentId segmentId)
+  public boolean cancelOperation(SegmentAction action, DataSegment segment)
+  {
+    final SegmentAction queuedAction = queuedSegments.get(segment);
+    return queuedAction == simplify(action)
+           && (queuedAction == SegmentAction.MOVE_FROM || peon.cancelOperation(segment))
+           && updateQueuedSegments(segment, queuedAction, false);
+  }
+
+  public boolean hasSegmentLoaded(SegmentId segmentId)
   {
     return server.getSegment(segmentId) != null;
   }
 
-  /**
-   * Checks if the server can load the given segment.
-   * <p>
-   * A load is possible only if the server meets all of the following criteria:
-   * <ul>
-   *   <li>is not being decommissioned</li>
-   *   <li>is not already serving the segment</li>
-   *   <li>is not performing any other action on the segment</li>
-   *   <li>has not already exceeded the load queue limit in this run</li>
-   *   <li>has available disk space</li>
-   * </ul>
-   */
-  public boolean canLoadSegment(DataSegment segment)
+  public boolean isRealtimeServer()
   {
-    return !isDecommissioning
-           && !isServingSegment(segment.getId())
-           && !isLoadingSegment(segment)
-           && (maxSegmentsInLoadQueue == 0 || maxSegmentsInLoadQueue > peon.getNumberOfSegmentsInQueue())
-           && getAvailableSize() >= segment.getSize();
+    return server.getType() == ServerType.REALTIME
+           || server.getType() == ServerType.INDEXER_EXECUTOR;
   }
 
-  @Override
-  public int compareTo(ServerHolder serverHolder)
+  private SegmentAction simplify(SegmentAction action)
   {
-    int result = Long.compare(getAvailableSize(), serverHolder.getAvailableSize());
-    if (result != 0) {
-      return result;
+    return action == SegmentAction.REPLICATE ? SegmentAction.LOAD : action;
+  }
+
+  private boolean updateQueuedSegments(DataSegment segment, SegmentAction action, boolean addToQueue)
+  {
+    if (addToQueue) {
+      queuedSegments.put(segment, action);
+    } else {
+      queuedSegments.remove(segment);
     }
 
-    result = server.getHost().compareTo(serverHolder.server.getHost());
-    if (result != 0) {
-      return result;
+    final long sizeDelta = addToQueue ? segment.getSize() : -segment.getSize();
+    if (action.isLoad()) {
+      sizeOfLoadingSegments += sizeDelta;
+    } else if (action == SegmentAction.DROP) {
+      sizeOfDroppingSegments += sizeDelta;
     }
 
-    result = server.getTier().compareTo(serverHolder.server.getTier());
-    if (result != 0) {
-      return result;
+    // Remove from projected if load is cancelled or drop is started, add otherwise
+    if (addToQueue ^ action.isLoad()) {
+      projectedSegments.remove(segment);
+    } else {
+      projectedSegments.add(segment);
     }
 
-    return server.getType().compareTo(serverHolder.server.getType());
+    return true;
+  }
```

Comment:
nit: the boolean parameter here feels really weird.  Why not just have 2 methods:

```
addQueuedSegment()
removeQueuedSegment()
```

Reply:
Yeah, the boolean here is stupid, it's always `true`, must have been doing some other things originally. With the recent changes, it would have become redundant. Can get rid of it without breaking up the method too.